contracts/actions/mcd/McdRepayComposite.sol

```diff
@@ -0,0 +1,172 @@
+    function _repay(RepayParams memory _repayParams) internal returns (bytes memory logData) {
+        assert(_repayParams.exchangeData.destAddr == DAI_ADDR);
+        uint256 collateral = getAllColl(IManager(MCD_MANAGER_ADDR), _repayParams.joinAddr, _repayParams.vaultId);
+        uint256 repayAmount = _repayParams.exchangeData.srcAmount > collateral ? collateral : _repayParams.exchangeData.srcAmount;
+        _repayParams.exchangeData.srcAmount = repayAmount;
+
+        // Withdraw collateral
+        {
+            uint256 frobAmount = convertTo18(_repayParams.joinAddr, repayAmount);
+            IManager(MCD_MANAGER_ADDR).frob(
+                _repayParams.vaultId,
+                -toPositiveInt(frobAmount),
+                0
+            );
+            IManager(MCD_MANAGER_ADDR).flux(_repayParams.vaultId, address(this), frobAmount);
+            IJoin(_repayParams.joinAddr).exit(address(this), repayAmount);
+        }
+
+        // Sell collateral asset for debt asset
+        (uint256 exchangedAmount, ) = _dfsSell(_repayParams.exchangeData, address(this), address(this), false);
+
+        // Take gas fee
+        uint256 paybackAmount = _takeFee(GasFeeTakerParams(
+            _repayParams.gasUsed,
+            DAI_ADDR,
+            exchangedAmount,
+            0
+        ));
+
+        // Payback debt
+        {
+            // if paybackAmount is higher than current debt, repay all debt and send remaining dai to proxy owner
+            (address urn, bytes32 ilk) = getUrnAndIlk(MCD_MANAGER_ADDR, _repayParams.vaultId);
+            uint256 debt = getAllDebt(address(vat), urn, urn, ilk);
+            if (paybackAmount > debt) {
+                DAI_ADDR.withdrawTokens(IDSProxy(address(this)).owner(), paybackAmount - debt);
+                paybackAmount = debt;
+            }
+
+            DAI_ADDR.approveToken(DAI_JOIN_ADDR, paybackAmount);
+            IDaiJoin(DAI_JOIN_ADDR).join(urn, paybackAmount);
+            uint256 daiVatBalance = vat.dai(urn);
+            int256 paybackAmountNormalized = normalizePaybackAmount(address(vat), daiVatBalance, urn, ilk);
+
+            IManager(MCD_MANAGER_ADDR).frob(
+                _repayParams.vaultId,
+                0,
+                paybackAmountNormalized
+            );
+        }
+
+        logData = abi.encode(
+            address(this),
+            repayAmount,
+            exchangedAmount,
+            paybackAmount
+        );
+    }
+
+    /// @inheritdoc ActionBase
+    function actionType()
+    public
+    pure
+    virtual override (ActionBase, DFSSell, GasFeeTaker)
+    returns (uint8) {
+        return uint8(ActionType.CUSTOM_ACTION);
+    }
+
+    function _parseCompositeParams(bytes memory _calldata) internal pure returns (RepayParams memory params) {
+        params = abi.decode(_calldata, (RepayParams));
+    }
+
+    /// @dev workaround for dfsSell expecting context to be proxy
+    function owner() external pure returns (address) {
+        return address(0);
```

Comment:
Can this field be 0x0 and 0x and other exchanges still working normally/tracking not changed?

Reply:
And isn't this call in the context of dsproxy already?

Reply:
Negligent use of copy paste, fixed.